### ISSUE TITLE
Default values for lat/long

### ISSUE THREAD

**Issue by caseyprovost** [rgeo/activerecord-postgis-adapter#344]
We are storing coordinates with a default right now with a schema entry like so:

```ruby
t.geography "coordinates", limit: {:srid=>4326, :type=>"st_point", :geographic=>true}, default: "POINT(0.0 0.0)"
```

While trying to upgrade to 7.1.0 we are getting the error below. If we remove the default definition though, everything works. Hoping someone can shed some light on how to tweak or change the default definition appropriately for the current version. I've yet to find anything in the docs or code to suggest how to get past this.

```
NoMethodError: undefined method `match?' for nil:NilClass
/bundle/ruby/2.6.0/gems/activerecord-postgis-adapter-7.1.0/lib/active_record/connection_adapters/postgis/oid/spatial.rb:47:in `parse_sql_type'
/bundle/ruby/2.6.0/gems/activerecord-postgis-adapter-7.1.0/lib/active_record/connection_adapters/postgis/schema_statements.rb:106:in `block (2 levels) in initialize_type_map'
/bundle/ruby/2.6.0/gems/activerecord-6.1.3/lib/active_record/type/hash_lookup_type_map.rb:20:in `perform_fetch'
/bundle/ruby/2.6.0/gems/activerecord-6.1.3/lib/active_record/type/type_map.rb:21:in `block in fetch'
/bundle/ruby/2.6.0/gems/concurrent-ruby-1.1.6/lib/concurrent-ruby/concurrent/map.rb:193:in `block in fetch_or_store'
/bundle/ruby/2.6.0/gems/concurrent-ruby-1.1.6/lib/concurrent-ruby/concurrent/map.rb:172:in `fetch'
/bundle/ruby/2.6.0/gems/concurrent-ruby-1.1.6/lib/concurrent-ruby/concurrent/map.rb:192:in `fetch_or_store'
/bundle/ruby/2.6.0/gems/activerecord-6.1.3/lib/active_record/type/type_map.rb:20:in `fetch'
/bundle/ruby/2.6.0/gems/activerecord-6.1.3/lib/active_record/type/type_map.rb:16:in `lookup'
/bundle/ruby/2.6.0/gems/activerecord-6.1.3/lib/active_record/type/hash_lookup_type_map.rb:7:in `block in alias_type'
/bundle/ruby/2.6.0/gems/activerecord-6.1.3/lib/active_record/type/hash_lookup_type_map.rb:20:in `perform_fetch'
/bundle/ruby/2.6.0/gems/activerecord-6.1.3/lib/active_record/type/type_map.rb:21:in `block in fetch'
/bundle/ruby/2.6.0/gems/concurrent-ruby-1.1.6/lib/concurrent-ruby/concurrent/map.rb:193:in `block in fetch_or_store'
/bundle/ruby/2.6.0/gems/concurrent-ruby-1.1.6/lib/concurrent-ruby/concurrent/map.rb:172:in `fetch'
/bundle/ruby/2.6.0/gems/concurrent-ruby-1.1.6/lib/concurrent-ruby/concurrent/map.rb:192:in `fetch_or_store'
/bundle/ruby/2.6.0/gems/activerecord-6.1.3/lib/active_record/type/type_map.rb:20:in `fetch'
/bundle/ruby/2.6.0/gems/activerecord-6.1.3/lib/active_record/type/type_map.rb:16:in `lookup'
/bundle/ruby/2.6.0/gems/activerecord-6.1.3/lib/active_record/connection_adapters/abstract/quoting.rb:214:in `lookup_cast_type'
/bundle/ruby/2.6.0/gems/activerecord-6.1.3/lib/active_record/connection_adapters/postgresql/quoting.rb:120:in `lookup_cast_type'
/bundle/ruby/2.6.0/gems/activerecord-6.1.3/lib/active_record/connection_adapters/abstract/quoting.rb:91:in `quote_default_expression'
/bundle/ruby/2.6.0/gems/activerecord-6.1.3/lib/active_record/connection_adapters/postgresql/quoting.rb:73:in `quote_default_expression'
/bundle/ruby/2.6.0/gems/activerecord-6.1.3/lib/active_record/connection_adapters/abstract/schema_creation.rb:16:in `quote_default_expression'
/bundle/ruby/2.6.0/gems/activerecord-6.1.3/lib/active_record/connection_adapters/abstract/schema_creation.rb:138:in `add_column_options!'
/bundle/ruby/2.6.0/gems/activerecord-6.1.3/lib/active_record/connection_adapters/postgresql/schema_creation.rb:64:in `add_column_options!'
/bundle/ruby/2.6.0/gems/activerecord-6.1.3/lib/active_record/connection_adapters/abstract/schema_creation.rb:34:in `visit_ColumnDefinition'
/bundle/ruby/2.6.0/gems/activerecord-6.1.3/lib/active_record/connection_adapters/abstract/schema_creation.rb:13:in `accept'
/bundle/ruby/2.6.0/gems/activerecord-6.1.3/lib/active_record/connection_adapters/abstract/schema_creation.rb:47:in `block in visit_TableDefinition'
/bundle/ruby/2.6.0/gems/activerecord-6.1.3/lib/active_record/connection_adapters/abstract/schema_creation.rb:47:in `map'
/bundle/ruby/2.6.0/gems/activerecord-6.1.3/lib/active_record/connection_adapters/abstract/schema_creation.rb:47:in `visit_TableDefinition'
/bundle/ruby/2.6.0/gems/activerecord-6.1.3/lib/active_record/connection_adapters/abstract/schema_creation.rb:13:in `accept'
/bundle/ruby/2.6.0/gems/activerecord-6.1.3/lib/active_record/connection_adapters/abstract/schema_statements.rb:322:in `create_table'
/bundle/ruby/2.6.0/gems/activerecord-6.1.3/lib/active_record/migration.rb:929:in `block in method_missing'
/bundle/ruby/2.6.0/gems/activerecord-6.1.3/lib/active_record/migration.rb:897:in `block in say_with_time'
/bundle/ruby/2.6.0/gems/benchmark-0.1.1/lib/benchmark.rb:293:in `measure'
/bundle/ruby/2.6.0/gems/activerecord-6.1.3/lib/active_record/migration.rb:897:in `say_with_time'
/bundle/ruby/2.6.0/gems/activerecord-6.1.3/lib/active_record/migration.rb:918:in `method_missing'
/bundle/ruby/2.6.0/bundler/gems/strong_migrations-dc1ed9f9321b/lib/strong_migrations/migration.rb:78:in `method_missing'
/app/db/schema.rb:474:in `block in <top (required)>'
/bundle/ruby/2.6.0/gems/activerecord-6.1.3/lib/active_record/schema.rb:50:in `instance_eval'
/bundle/ruby/2.6.0/gems/activerecord-6.1.3/lib/active_record/schema.rb:50:in `define'
/bundle/ruby/2.6.0/gems/activerecord-6.1.3/lib/active_record/schema.rb:46:in `define'
/app/db/schema.rb:13:in `<top (required)>'
/bundle/ruby/2.6.0/gems/bootsnap-1.4.6/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:55:in `load'
/bundle/ruby/2.6.0/gems/bootsnap-1.4.6/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:55:in `load'
/bundle/ruby/2.6.0/gems/activerecord-6.1.3/lib/active_record/tasks/database_tasks.rb:326:in `load_schema'
/bundle/ruby/2.6.0/gems/activerecord-6.1.3/lib/active_record/tasks/database_tasks.rb:435:in `block in load_schema_current'
/bundle/ruby/2.6.0/gems/activerecord-6.1.3/lib/active_record/tasks/database_tasks.rb:506:in `block (2 levels) in each_current_configuration'
/bundle/ruby/2.6.0/gems/activerecord-6.1.3/lib/active_record/tasks/database_tasks.rb:503:in `each'
/bundle/ruby/2.6.0/gems/activerecord-6.1.3/lib/active_record/tasks/database_tasks.rb:503:in `block in each_current_configuration'
/bundle/ruby/2.6.0/gems/activerecord-6.1.3/lib/active_record/tasks/database_tasks.rb:502:in `each'
/bundle/ruby/2.6.0/gems/activerecord-6.1.3/lib/active_record/tasks/database_tasks.rb:502:in `each_current_configuration'
/bundle/ruby/2.6.0/gems/activerecord-6.1.3/lib/active_record/tasks/database_tasks.rb:434:in `load_schema_current'
/bundle/ruby/2.6.0/gems/activerecord-6.1.3/lib/active_record/railties/databases.rake:461:in `block (3 levels) in <top (required)>'
/bundle/ruby/2.6.0/gems/rake-13.0.3/lib/rake/task.rb:281:in `block in execute'
/bundle/ruby/2.6.0/gems/rake-13.0.3/lib/rake/task.rb:281:in `each'
/bundle/ruby/2.6.0/gems/rake-13.0.3/lib/rake/task.rb:281:in `execute'
/bundle/ruby/2.6.0/gems/rake-13.0.3/lib/rake/task.rb:219:in `block in invoke_with_call_chain'
/usr/local/lib/ruby/2.6.0/monitor.rb:235:in `mon_synchronize'
/bundle/ruby/2.6.0/gems/rake-13.0.3/lib/rake/task.rb:199:in `invoke_with_call_chain'
/bundle/ruby/2.6.0/gems/rake-13.0.3/lib/rake/task.rb:243:in `block in invoke_prerequisites'
/bundle/ruby/2.6.0/gems/rake-13.0.3/lib/rake/task.rb:241:in `each'
/bundle/ruby/2.6.0/gems/rake-13.0.3/lib/rake/task.rb:241:in `invoke_prerequisites'
/bundle/ruby/2.6.0/gems/rake-13.0.3/lib/rake/task.rb:218:in `block in invoke_with_call_chain'
/usr/local/lib/ruby/2.6.0/monitor.rb:235:in `mon_synchronize'
/bundle/ruby/2.6.0/gems/rake-13.0.3/lib/rake/task.rb:199:in `invoke_with_call_chain'
/bundle/ruby/2.6.0/gems/rake-13.0.3/lib/rake/task.rb:188:in `invoke'
/bundle/ruby/2.6.0/gems/rake-13.0.3/lib/rake/application.rb:160:in `invoke_task'
/bundle/ruby/2.6.0/gems/rake-13.0.3/lib/rake/application.rb:116:in `block (2 levels) in top_level'
/bundle/ruby/2.6.0/gems/rake-13.0.3/lib/rake/application.rb:116:in `each'
/bundle/ruby/2.6.0/gems/rake-13.0.3/lib/rake/application.rb:116:in `block in top_level'
/bundle/ruby/2.6.0/gems/rake-13.0.3/lib/rake/application.rb:125:in `run_with_threads'
/bundle/ruby/2.6.0/gems/rake-13.0.3/lib/rake/application.rb:110:in `top_level'
/bundle/ruby/2.6.0/gems/railties-6.1.3/lib/rails/commands/rake/rake_command.rb:24:in `block (2 levels) in perform'
/bundle/ruby/2.6.0/gems/rake-13.0.3/lib/rake/application.rb:186:in `standard_exception_handling'
/bundle/ruby/2.6.0/gems/railties-6.1.3/lib/rails/commands/rake/rake_command.rb:24:in `block in perform'
/bundle/ruby/2.6.0/gems/rake-13.0.3/lib/rake/rake_module.rb:59:in `with_application'
/bundle/ruby/2.6.0/gems/railties-6.1.3/lib/rails/commands/rake/rake_command.rb:18:in `perform'
/bundle/ruby/2.6.0/gems/railties-6.1.3/lib/rails/command.rb:52:in `invoke'
/bundle/ruby/2.6.0/gems/railties-6.1.3/lib/rails/commands.rb:18:in `<top (required)>'
/bundle/ruby/2.6.0/gems/bootsnap-1.4.6/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:23:in `require'
/bundle/ruby/2.6.0/gems/bootsnap-1.4.6/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:23:in `block in require_with_bootsnap_lfi'
/bundle/ruby/2.6.0/gems/bootsnap-1.4.6/lib/bootsnap/load_path_cache/loaded_features_index.rb:92:in `register'
/bundle/ruby/2.6.0/gems/bootsnap-1.4.6/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:22:in `require_with_bootsnap_lfi'
/bundle/ruby/2.6.0/gems/bootsnap-1.4.6/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:31:in `require'
/bundle/ruby/2.6.0/gems/activesupport-6.1.3/lib/active_support/dependencies.rb:332:in `block in require'
/bundle/ruby/2.6.0/gems/activesupport-6.1.3/lib/active_support/dependencies.rb:299:in `load_dependency'
/bundle/ruby/2.6.0/gems/activesupport-6.1.3/lib/active_support/dependencies.rb:332:in `require'
bin/rails:4:in `<main>'
Tasks: TOP => db:setup => db:schema:load
```

**Comment by keithdoggett** [comment]
I was able to look at this for a few minutes. The issue is that the default value has to be serialized by the proper spatial typecaster, but ActiveRecord does not send the rest of the column information over with it. Normally in the `initialize_type_map` block, we would get the SQL definition of the column, but `sql_type` is `nil` in this case.

https://github.com/rgeo/activerecord-postgis-adapter/blob/f7108d4456bf36b82b8340f69ab19aed976bb073/lib/active_record/connection_adapters/postgis/schema_statements.rb#L88-L109

There could be a potential workaround by checking if `sql_type` is `nil` and using the first argument of 3 in the block to see if it is a geography or geometry and making some assumptions based on that, but it is not robust and would probably produce erroneous values if the SRID is not 0 or 4326.

In older versions how did the default value appear in your schema? Was it a binary format (`0101000020E610000000000000000000000000000000000000`) or WKT (`POINT (0.0 0.0)`)?

I'm not sure how this was working before since it seems like it would just run into the same issue.

**Comment by caseyprovost** [comment]
@keithdoggett I am not entirely sure the history of how these values were set. In the schema they are set to "POINT(0.0 0.0)", but I suppose that could have been manual intervention. I know when we run migrations those default values look more like `<RGeo::Point`. I forget the exact class reference, but it does break the schema unless transformed back into the POINT value we have now.

Do we maybe just need to update the ReadMe to mention default value issues?

**Comment by keithdoggett** [comment]
@caseyprovost ok thanks for that info! I think the solution is probably to override part of the default value flow for spatial columns. I'm going to take another look at this soon, but if it becomes a bigger issue, I'll add a note to the readme.

**Comment by keithdoggett** [comment]
@caseyprovost  sorry about the delay on this. I've opened up #347 which should fix it, but if you're able to test that branch yourself that would be very helpful.

**Comment by caseyprovost** [comment]
Thanks @keithdoggett !! I worked around it by removing the DB defaults and setting them programmatically in the app. But I’ll def run an earlier commit and test this for you

**Comment by keithdoggett** [comment]
Fix released in versions 7.1.1 and 6.0.3
